src/wallet/wallet.cpp

```diff
@@ -3900,6 +3928,39 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
             pindex = chainActive.Next(pindex);
         }
 
+        // Attempt to rewind the orchard wallet to the rescan point if the wallet has any
+        // checkpoints. Note data will be restored by the calls to AddToWalletIfInvolvingMe,
+        // and then the call to `ChainTipAdded` that later occurs for each block will restore
+        // the witness data that is being removed in the rewind here.
+        auto nu5_height = chainParams.GetConsensus().GetActivationHeight(Consensus::UPGRADE_NU5);
+        bool performOrchardWalletUpdates{false};
+        if (orchardWallet.IsCheckpointed()) {
+            // We have a checkpoint, so attempt to rewind the Orchard wallet at most as
+            // far as the NU5 activation block.
+            // If there's no activation height, we shouldn't have a checkpoint already,
+            // and this is a programming error.
+            assert(nu5_height.has_value());
+            // Only attempt to perform scans for Orchard during wallet initialization,
+            // since we do not support Orchard key import.
+            if (isInitScan) {
+                if (orchardWallet.Rewind(std::max(nu5_height.value(), pindex->nHeight))) {
+                    // rewind was successful or a no-op, so perform Orchard wallet updates
+                    performOrchardWalletUpdates = true;
+                } else {
+                    // Orchard witnesses will not be able to be correctly updated, because we
+                    // can't rewind far enough. This is an unrecoverable failure; it means that we
+                    // can't get back to a valid wallet state without resetting the wallet all
+                    // the way back to NU5 activation.
+                    throw std::runtime_error("CWallet::ScanForWalletTransactions() : Orchard wallet needs to be resynced from NU5 activation.");
```

Comment:
More descriptive error for what the user needs to do (rerun with -rescan i guess?)